fix: consider previous Leave Applications while enforcing Maximum Consecutive Leaves Allowed

krantheman opened the pull request:
### Issue

Closes: https://github.com/frappe/hrms/issues/1002

Currently, separate consecutive Leave Applications are not checked by the **Maximum Consecutive Leaves Allowed** limit entered in the Leave Type DocType. This PR fixes that.

### Example

- **Maximum Consecutive Leaves Allowed** set to 4

![image](https://github.com/frappe/hrms/assets/61287991/67bc2062-4fe6-43e6-a4f5-06684b09d850)

- Two Leave Applications created spanning 4 days

![image](https://github.com/frappe/hrms/assets/61287991/30e3dfd8-7932-453f-a555-10cceacb99a5)

- Details filled in for a third Leave Application that would constitute the fifth consecutive leave of this Leave Type

![image](https://github.com/frappe/hrms/assets/61287991/b7924c68-9cf5-49e4-a64d-29bc216d3fa6)

- Action is forbidden as this violates the **Maximum Consecutive Leaves Allowed** limit that has been set

![image](https://github.com/frappe/hrms/assets/61287991/033c1ef1-cfb7-4b16-bdc8-de9693288acf)


`no-docs`

krantheman committed:
fix: consider previous Leave Applications while enforcing Maximum Consecutive Leaves Allowed

krantheman committed:
feat: better error msg

krantheman committed:
test: add test for first_from_date and last_to_date

krantheman committed:
chore: convert map to list comprehension

ruchamahabal commented on hrms/hr/doctype/leave_application/leave_application.py:
```diff
@@ -480,10 +481,70 @@ def get_total_leaves_on_half_day(self):
+			self.employee, self.leave_type, first_from_date, last_to_date
+		)
+		if total_consecutive_leaves > cint(max_days):
+			leave_applications = frappe.get_list(
```
```suggestion
			leave_applications = frappe.get_all(
```
We don't need permission sensitive query here

ruchamahabal commented on hrms/hr/doctype/leave_application/leave_application.py:
```diff
@@ -480,10 +481,70 @@ def get_total_leaves_on_half_day(self):
+				},
+				pluck="name",
+			)
+			if len(leave_applications) > 0:
```
```suggestion
			if leave_applications:
```
non-empty lists are truthy

ruchamahabal commented on hrms/hr/doctype/leave_application/leave_application.py:
```diff
@@ -480,10 +481,70 @@ def get_total_leaves_on_half_day(self):
+					[get_link_to_form("Leave Application", x) for x in leave_applications]
+				)
+				frappe.throw(
+					_("Cannot have more that {0} consecutive leaves of type {1}. Reference: {2}").format(
```
I think the original message was more user friendly

Leave of type {} cannot be longer than
And then mention the references

krantheman committed:
fix: permission sensitivity

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
chore: remove redundant code

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
chore: better error msg

krantheman committed:
fix: translations

ruchamahabal committed:
refactor: get consecutive leave details

- simplify getting dates recursively

- avoid extra db call to fetch applications between dates

- remove unnecessary list comprehension for generator function

ruchamahabal committed:
test: assert consecutive leaves validation

ruchamahabal committed:
fix: dont consider holidays while checking for consecutive leaves

ruchamahabal reviewed (commented):
@krantheman made some changes

1. you don't need to pass list comprehensions to functions like join that accept a generator

```python
", ".join([get_link_to_form("Leave Application", x) for x in leave_applications])
```
changed to

```python
", ".join(get_link_to_form("Leave Application", x) for x in leave_applications)
```
2. Avoided extra DB call to fetch leave applications with consecutive leaves - by maintaining a set to add docnames while getting first from date & last to date

3. Not considering holidays in between consecutive leaves, I don't think that really counts as consecutive leaves. Also, there are more edge cases to that scenario. Currently "Include holidays within leaves as leaves" is also applicable for only a single leave application so until that changes, we should use uniform behavior here too.

4. In tests, asserting validation error instead of checking the first & last dates. As far as possible, try testing for output rather than implementation (or both). Implementation (getting first & last dates) can change but the effect (throwing validation error) remains the same so your tests should validate that.
